### PR TITLE
ibAlluoV2, LiquidityBufferV2 (works for ETH) and USDAdaptor.

### PR DESCRIPTION
**Pull request by pentatonictritones**
Hi, this is the PR for #19 #30 .

When deploying for ETH, just set primary token to ETH. Should be very straightforward.

When deploying for USD, see how I've done it in the test file. Deploy ibAlluo, Liquidity Buffer and USD Adaptor. Set variables correctly as in the test file. Then all good to go!

Changes are:

I had to make some small changes to ibAlluo to make sure it is more generalisable and forward compatible. In particular:
1. function deposit():   I changed the interfaces used, moved transfer logic to the liquidity buffer so that it is consistent (see discussion above).

2. function withdraw(): Changed input _amount param to amount in stablecoins (makes more sense this way, consistent + intuitive for frontend work), changed order of transactions around so that it is more futureproof for adaptors and different tokens

For tests, I adapted tests for ibAlluo.ts into IbAlluoV2+BufferV2+USDAdaptor.ts. All tests pass.

I also wrote two separate test files: 1 is just to test the ETH case (when you have no adaptor, funds are just kept in the contract and multisig admin can take funds out if they want) and the other is to test the USD adaptor. Everything passes. I tested all the new logic (through different if statements in deposit/withdrawal.... etc)

### CODE REVIEW

File: contracts/Farming/IbAlluo.sol [hunk 1]
```diff
@@ -254,7 +254,29 @@ contract IbAlluo is
...
+    function migrateStep1(address _oldContract, address[] memory _users) external onlyRole(DEFAULT_ADMIN_ROLE){
+        for(uint i = 0; i < _users.length; i++){
+            uint256 oldBalance = AlluoERC20Upgradable(_oldContract).balanceOf(_users[i]);
+            _mint(_users[i], oldBalance);
+        }
+    }
+
+    /// @notice migrates by setting new interest variables.
+
+    function migrateStep2() external onlyRole(DEFAULT_ADMIN_ROLE){
```
Comment (alexanderem49) [inline contracts/Farming/IbAlluo.sol:269]: Why is this extracted into different function?

Comment (pentatonictritones) [reply contracts/Farming/IbAlluo.sol:269]: Hi, this snippet was written by the Alluo team - I believe to simplify the initial migration process from the old system to the index system and to make sure there weren't any errors when claiming balances.

File: contracts/Farming/LiquidityBufferUSDAdaptor.sol [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+     * NOTE: Input _amount is 10**18. Make sure to standardise it for the token (10**18 DAI, 10**6 USDC/USDT)
+     ** @param _token Address of the token input
+     ** @param _primaryToken Address of the primary Token to be sent to the wallet (USDC here)
+     ** @param _wallet Address of the multisig wallet 
+     ** @param _amount Amount of tokens in 10**18
+     ** @param _pool Address of liquidity pool to be used.
+     ** @param slippage Slippage specified from Liquidity buffer
+     */
+    function enterAdaptor(address _token, address _primaryToken, address _wallet, uint256 _amount, address _pool, uint256 slippage) external returns (uint256) {
+        if (_token == 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063) {
```
Comment (zayk00v) [inline contracts/Farming/LiquidityBufferUSDAdaptor.sol:21]: Can you please change these addresses for constants?

"The compiler does not reserve a storage slot for these variables, and every occurrence is replaced by the respective value."

[https://docs.soliditylang.org/en/v0.8.11/contracts.html#constant-and-immutable-state-variables](url)

Comment (zayk00v) [inline contracts/Farming/LiquidityBufferUSDAdaptor.sol:21]: Can you please change these addresses for constants?

"The compiler does not reserve a storage slot for these variables, and every occurrence is replaced by the respective value."

[https://docs.soliditylang.org/en/v0.8.11/contracts.html#constant-and-immutable-state-variables](url)

**Pull request closed without merging**